src/Google.Cloud.Diagnostics.Debug/Google.Cloud.Diagnostics.Debug/LoggingClient.cs

```diff
@@ -0,0 +1,133 @@
+                {StackdriverBreakpoint.Types.LogLevel.Error, LogSeverity.Error }
+        };
+
+        private readonly LoggingServiceV2Client _logClient;
+        private readonly AgentOptions _options;
+        private readonly LogName _logName;
+
+        internal LoggingClient(AgentOptions options, LoggingServiceV2Client loggingClient = null)
+        {
+            _logClient = loggingClient ?? LoggingServiceV2Client.Create();
+            _options = GaxPreconditions.CheckNotNull(options, nameof(options));
+            _logName = new LogName(_options.ProjectId, _options.LogName ??
+                $"{_options.Module}-{_options.Version}-debug-log");
+        }
+        
+        /// <summary>
+        /// Substitutes the log message format in breakpoint and writes
+        /// the result as a log entry to the log _logName.
+        /// </summary>
+        /// <returns></returns>
+        public WriteLogEntriesResponse WriteLogEntry(StackdriverBreakpoint breakpoint)
+        {
+            LogEntry logEntry = new LogEntry
+            {
+                LogName = _logName.ToString(),
+                Severity = _logSeverityConversion[breakpoint.LogLevel],
+                TextPayload = SubstituteLogMessageFormat(breakpoint.LogMessageFormat, breakpoint.EvaluatedExpressions.ToList())
+            };
+            MonitoredResource resource = new MonitoredResource { Type = "global" };
+            return _logClient.WriteLogEntries(LogNameOneof.From(_logName), resource, null, new[] { logEntry });
+        }
+
+        /// <summary>
+        /// Substitutes the $0, $1, etc. in messageFormat with expressions
+        /// from evaluatedExpressions.
+        /// </summary>
+        /// <returns></returns>
+        private string SubstituteLogMessageFormat(string messageFormat, List<Debugger.V2.Variable> evaluatedExpressions)
+        {
+            string result = "";
+            int offset = 0;
+            int i = 0;
+            while (i < messageFormat.Length)
+            {
+                if (messageFormat[i] != '$')
+                {
+                    i += 1;
+                    continue;
+                }
+
+                // We encounter a $.
+                result += messageFormat.Substring(offset, i - offset);
+
+                char nextChar = messageFormat[i + 1];
+                if (nextChar == '$')
+                {
+                    // Escape the current $.
+                    result += '$';
+                    i += 2;
+                    offset = i;
+                    continue;
+                }
+
+                if (!Char.IsDigit(nextChar))
+                {
+                    offset = i;
+                    i += 1;
+                    continue;
+                }
+
+                // We have a number followed a $. Time to substitute!
+                string currentNumberString = "";
+                i += 1;
+                while (i < messageFormat.Length && Char.IsDigit(messageFormat[i]))
```

Comment:
All of the agents (and the UI IIRC) assume we never have more than 10 (0-9) expressions.  So feel free to ignore this to make things less complicated.